### PR TITLE
add py312 support - v6

### PR DESCRIPTION
**Pull request by pacrob**
### What was wrong?

Support py312 in v6. Bumped `flake8` version to one that supports py312, fixed a couple errors that came up.

### Todo:

- [x] Clean up commit history
- [x] Add or update documentation related to these changes
- [x] Add entry to the [release notes](https://github.com/ethereum/web3.py/blob/main/newsfragments/README.md)

#### Cute Animal Picture

image

### CODE REVIEW

File: web3/tools/pytest_ethereum/deployer.py [hunk 1]
```diff
@@ -1,8 +1,8 @@
 from typing import (
     Any,
     Callable,
-    Dict,
 )
+from typing import Dict  # noqa: F401
```
Comment (reedsa) [inline web3/tools/pytest_ethereum/deployer.py]: Was there a limit your for dict? Weird that it imports typing twice

Comment (pacrob) [reply web3/tools/pytest_ethereum/deployer.py]: Not sure. I put the `noqa` line after it in the original single import and something in linting split it out. Maybe isort?

Comment (kclowes) [reply web3/tools/pytest_ethereum/deployer.py]: I think you can say: `self.strategies: Dict[str, Callable[[Package], Package]] = {}` on line 30 below and get rid of this weird import stuff

**Review by kclowes — APPROVED** [review]
Sweet!

**Review by fselmo — APPROVED** [review]
lgtm!

**Pull request merged**